### PR TITLE
Relay statement parameters to doctrine dbal

### PR DESCRIPTION
**Pull request by ausi**
This rewires our legacy database statements to doctrine to use real prepared statements.

AFAIK, we only supported `?` for parameters and `%s` for `INSERT`/`UPDATE` sets, is this correct?

### CODE REVIEW

**Comment by aschempp** [comment]
Didn't we have issues with the `set` method when we previouly tried that?

**Comment by ausi** [comment]
> Didn't we have issues with the `set` method when we previouly tried that?

I could only find https://github.com/contao/core-bundle/pull/182 and https://github.com/contao/core-bundle/pull/31 and I didn’t see problems with `set()` there.

**Comment by aschempp** [comment]
Whats with the condition in the `set()` method that is not supported (anymore) with this PR? Was `set()` previously supported for other statements or was the message just missing?

**Comment by ausi** [comment]
> Was `set()` previously supported for other statements or was the message just missing?

See https://github.com/contao/contao/blob/722d397e3429e3fc4a8d05263fe2f834aa8f2122/core-bundle/src/Resources/contao/library/Contao/Database/Statement.php#L167

File: core-bundle/tests/Contao/Database/StatementTest.php [hunk 1]
```diff
@@ -0,0 +1,346 @@
...
+
+        $this->expectDeprecation('%sUsing "%s" with an array parameter has been deprecated%s');
+
+        $statement->execute($params ?? []);
+    }
+
+    /**
+     * @dataProvider getQueriesWithParametersAndSets
+     */
+    /*
```
Comment (leofeyer) [inline core-bundle/tests/Contao/Database/StatementTest.php]: Since this is commented out, I guess it can be removed?

**Comment by leofeyer** [comment]
> AFAIK, we only supported `?` for parameters and `%s` for `INSERT`/`UPDATE` sets, is this correct?

Yes.

File: core-bundle/tests/Contao/Database/StatementTest.php (hunk 1 above)
Comment (ausi) [reply core-bundle/tests/Contao/Database/StatementTest.php]: I’d wait for the reviews. This shows how the old implementation worked. Once approved we should remove it before the PR gets merged.

File: core-bundle/src/Resources/contao/library/Contao/Database/Statement.php [hunk 2]
```diff
@@ -270,8 +263,44 @@ public function query($strQuery='')
...
+					!\is_string($varParam)
+					&& !\is_bool($varParam)
+					&& !\is_float($varParam)
+					&& !\is_int($varParam)
+					&& $varParam !== null
+				) {
+					return serialize($varParam);
+				}
+
+				return $varParam;
```
Comment (m-vo) [inline core-bundle/src/Resources/contao/library/Contao/Database/Statement.php]: Maybe better comprehensible when de-morganified? :slightly_smiling_face: 
```suggestion
				if (
					\is_string($varParam)
					|| \is_bool($varParam)
					|| \is_float($varParam)
					|| \is_int($varParam)
					|| $varParam === null
				) {
					return $varParam;
				}

				return serialize($varParam);
```

**Review by leofeyer — APPROVED** [review]

**Comment by leofeyer** [comment]
Thank you @ausi.

**Pull request merged**